### PR TITLE
Add tests showing SharedCell drops its contents

### PR DESCRIPTION
**Pull request by yamadapc**
This commits demos what I understand as a bug in `SharedCell`. It'll
drop its contents (rather than hold the shared reference).

### CODE REVIEW

File: src/shared_cell.rs [hunk 1]
```diff
@@ -145,3 +145,30 @@ impl<T> Drop for SharedCell<T> {
...
+            fn drop(&mut self) {
+                let mut lock = self.0.lock().unwrap();
+                *lock = true;
+            }
+        }
+        let mut collector = Collector::new();
+        let has_dropped = Arc::new(Mutex::new(false));
+        let owned = SharedCell::new(Shared::new(&collector.handle(), Test(has_dropped.clone())));
+        collector.collect();
+        assert_eq!(*has_dropped.lock().unwrap().deref(), false);
```
Comment (yamadapc) [inline src/shared_cell.rs:171]: This assertion fails, while it shouldn't (?)

**Comment by micahrj** [comment]
You're right, this is a bug. Thanks for finding this. It's happening because `SharedCell::new()` is allowing its argument to be dropped rather than calling `mem::forget`. I'll push a fix.

**Comment by yamadapc** [comment]
Does it make sense to add tests anyway? Maybe I can disable `no_std` for tests only?

**Comment by micahrj** [comment]
I found two other bugs in the implementation of `SharedCell`, so I wrote my own test that checks for all of them. Again, I really appreciate you finding this bug and writing a test case for it! :)

For the future, it is actually possible to enable `std` for tests only, by adding the line `extern crate std` to a test module: https://github.com/glowcoil/basedrop/blob/34d82d0a1f6031750c9062505c65c265f0ee22ff/src/collector.rs#L348

**Pull request closed without merging**